### ISSUE TITLE
SDL_DropEvent loses sourceApplication string when passing from AppDelegate to SDL.

### ISSUE THREAD

**Issue by SDLBugzilla** [libsdl-org/SDL#2024]
# This bug report was migrated from our old Bugzilla tracker.

These attachments are available in the static archive:

* ~~[SDL_DropEvent srcApplication Patch. (drop_event_srcApp.patch, text/plain, 2015-11-15 22:23:44 +0000, 4530 bytes)](https://bugzilla.libsdl.org/attachment.cgi?id=2303)~~
* ~~[SDL_DropEvent srcApplication Patch + PC (drop_event_srcApp.patch, text/plain, 2015-11-21 06:21:00 +0000, 7770 bytes)](https://bugzilla.libsdl.org/attachment.cgi?id=2311)~~
* ~~[Updated Patch to add sourceApplication to DropEvent: Fixes comments. (drop_event_srcApp.patch, text/plain, 2015-11-27 04:56:54 +0000, 7779 bytes)](https://bugzilla.libsdl.org/attachment.cgi?id=2314)~~
* [Updated patch -- merged in changes since 2016-02, including new SendDropText and SendDropComplete. (patch.diff, text/plain, 2017-06-15 04:38:48 +0000, 7770 bytes)](https://bugzilla.libsdl.org/attachment.cgi?id=2762)

**Reported in version:** HG 2.1
**Reported for operating system, platform:** iOS (All), iPhone/iPod touch

# Comments on the original bug report:

On 2015-11-15 22:23:44 +0000, Nathan Daly wrote:

> Created attachment 2303
> SDL_DropEvent srcApplication Patch.
> 
> SDL_DropEvent only contains a `char *file` field, but is used to convert the iOS handleURL function into SDL, which also provides a string for the sourceApplication.
> 
> Some 3rd party libraries depend on the sourceApplication field being provided, but it is not currently available in SDL. (For me, I was trying to SignIn to GooglePlayGames services, but it fails unless you provide the sourceApplication string.)
> 
> 
> I've provided a patch that also pipes this string along to the SDL_DropEvent event structure.
> 
> 
> When I first wrote this code, I was building 2.0.3, so I hadn't yet seen there is Android support for this Event now too, added in Bug 2762.
> 
> Android doesn't provide information about the calling Activity when handling an Intent (as far as I can tell, anyway), so I just set this string field to NULL on Android. Is that acceptable?
> 
> 
> 
> Thanks! (This is my first time submitting to SDL, or generating a mercurial patch for that matter, so please let me know if I need to change anything!) ^.^
> ~Nathan

On 2015-11-20 08:46:47 +0000, Nathan Daly wrote:

> I've added Philipp and re-assigned to Ryan since I saw you two were the ones who originally responded to Bug 2762.
> 
> Thanks! :)

On 2015-11-20 20:40:23 +0000, Philipp Wiesemann wrote:

> Thank you for working on a patch.
> 
> The calls to SDL_SendDropFile() also need to be updated (with NULL) for the desktop operating systems because this function now has two parameters.

On 2015-11-21 06:21:00 +0000, Nathan Daly wrote:

> Created attachment 2311
> SDL_DropEvent srcApplication Patch + PC
> 
> Adds NULL to the calls in the Mac & Windows implementations in addition to the Android and iOS I had already done.

On 2015-11-21 06:22:31 +0000, Nathan Daly wrote:

> Thanks Phillipp for pointing that out! For whatever reason I hadn't seen the calls from the desktop code in my source files when I first wrote the patch.
> 
> I appreciate your quick response! :)

On 2015-11-26 21:00:33 +0000, Philipp Wiesemann wrote:

> Thank you for updating the patch.
> 
> The documentation at SDL_DropEvent is now a bit misleading. Though the field will (currently) be NULL on Android, it may also be NULL on some the other systems.
> 
> Maybe it should just be documented that it is NULL if the source is not known. This would also point users to always delete the string because they can not assume they need not (e.g. for only being on Android). Which again would improve upward compatibility of their programs a bit (with this feature creating a memory leak for older programs anyway :).

On 2015-11-27 04:54:42 +0000, Nathan Daly wrote:

> Oh, good catch. I'm sorry I didn't update the comments too.
> 
> 
> Oh man, you're right though: I hadn't considered that it will create a memory leak for all older programs using SDL_DropEvents. Wow, writing a C library is so hard! How do you ever prevent this problem when updating an interface? (I'm coming from a mostly C++ world.)
> 
> Is the right thing to instead introduce like a new SDL_DropEventWithSource and deprecate SDL_DropEvent? And then in the next release remove the old one? That way people will get a compile error and be forced to fix it, at least, and in the process of fixing it they can read about the new field. How do you normally proceed with something like this?
> 
> 
> I'll upload the corrected comments, but let me know if there is a better overall approach (maybe like I described above).
> 
> Thanks!

On 2015-11-27 04:56:54 +0000, Nathan Daly wrote:

> Created attachment 2314
> Updated Patch to add sourceApplication to DropEvent: Fixes comments.
> 
> This patch fixes the comment to note that sourceApplication may be null if source is unknown.

On 2015-11-27 05:08:57 +0000, Nathan Daly wrote:

> Comment on attachment 2314
> Updated Patch to add sourceApplication to DropEvent: Fixes comments.
> 
> Also, one more question, is there any style guide regarding 80char limits? The only lines that I modified that go over 80 chars are the actual SDL_SendDropFile() calls in /src/video/uikit/SDL_uikitappdelegate.m. Should I split those lines at the comma? I wasn't sure because the function signature is like 143 characters.

On 2015-11-28 21:16:07 +0000, Philipp Wiesemann wrote:

> (In reply to Nathan Daly from comment # 6)
> > How do you normally proceed with something like this?
> 
> Backward incompatible changes are mostly prevented by adding a hint to explicitly activate a new behavior. But this might not be useful here because some systems add the drop event before the hint could be set from the application.

On 2015-12-26 02:18:33 +0000, Nathan Daly wrote:

> Friendly Ping. I think barring any other decision about how to handle the memory leak (such as adding a new struct gated by a compile-time flag and deprecating the old one until it can be deleted in the future, forcing a compile-error) this is all set to be merged in.
> 
> Thanks! :)
> 
> Please let me know if you would like me to do anything different regarding the new memory leak. :)

On 2015-12-29 07:25:25 +0000, Ryan C. Gordon wrote:

> I think we're going to hold off on this for 2.0.4 (which hopefully is about to be finished) and deal with it in 2.0.5.
> 
> --ryan.

On 2015-12-29 22:39:13 +0000, Nathan Daly wrote:

> Okay sounds great. Thanks Ryan!

On 2016-02-04 07:27:34 +0000, Nathan Daly wrote:

> (In reply to Ryan C. Gordon from comment # 11)
> > I think we're going to hold off on this for 2.0.4 (which hopefully is about
> > to be finished) and deal with it in 2.0.5.
> > 
> > --ryan.
> 
> Looks like 2.0.4 was released successfully! :) Congrats.
> 
> Did you still want to look at merging this in? Thanks!

On 2016-02-19 04:58:17 +0000, Nathan Daly wrote:

> Friendly ping! :)

On 2016-10-01 20:26:10 +0000, Sam Lantinga wrote:

> Can you update this patch to work with the current SDL source?

On 2017-06-15 04:38:48 +0000, Nathan Daly wrote:

> Created attachment 2762
> Updated patch -- merged in changes since 2016-02, including new SendDropText and SendDropComplete.
> 
> I've updated the patch to be compatible with the new code, including the new functions SDL_SendDropText SDL_SendDropComplete.
> 
> There is one remaining question I have, and I left it as a TODO in the code:
> 
> What should we do for those two new functions. Should I also go through and add the sourceApplication to those? i don't know where it would come from, or if it's even meaningful (for now I have it sending NULL for those). If we do decide to do this, can we follow it up with a later patch?
> 
> Thanks!
> 
> Sorry it took me so long to get back to this!! :)

On 2017-08-12 19:18:50 +0000, Sam Lantinga wrote:

> We can make this kind of API change in SDL 2.1.
> 
> We could also potentially keep a string pool of application sources in the iOS video driver and have the event point at those strings, so it wouldn't need to be freed.
> 
> Do you know how long those strings are? If they're relatively short, we can put them into the event directly so we don't need a separate string pointer.

On 2017-08-14 18:27:39 +0000, Nathan Daly wrote:

> No sorry, I don't have any idea how long these strings are. It's been years since I was playing with this. :/
> 
> The pool could also make sense.
